File: addTwoNumbers.cpp

// leetcode.com 445. Add Two Numbers II

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    // copy the code from codefights' solution
    ListNode* addTwoNumbers(ListNode* a, ListNode* b) {
        vector<int> v1, v2;

        while (a)
        {
            v1.insert(v1.begin(), a->val);
            a = a->next;
        }

        while (b)
        {
            v2.insert(v2.begin(), b->val);
            b = b->next;
        }

        vector<int> res;
        int i, carry = 0, s = 0;

        for (i = 0; i < v1.size() && i < v2.size(); i++)
        {
            s = (v1[i] + v2[i] + carry) % 10;
            res.push_back(s);
            carry = (v1[i] + v2[i] + carry) / 10;
        }

        while (i < v1.size())
        {
            s = (v1[i] + carry) % 10;
            res.push_back(s);
            carry = (v1[i] + carry) / 10;
            ++i;
        }

        while (i < v2.size())
        {
            s = (v2[i] + carry) % 10;
            res.push_back(s);
            carry = (v2[i] + carry) / 10;
            ++i;
        }

        if (carry)
            res.push_back(carry);

        ListNode *header = new ListNode(-1);

        for (int i = 0; i < res.size(); i++)
        {
            ListNode *node = new ListNode(res[i]);
            node->next = header->next;
            header->next = node;
        }

        return header->next;        
    }
};

File: lintcode/src/main/java/com/vkeonline/lintcode/p100/AddTwoNumbers.java
package com.vkeonline.lintcode.p100;

import com.vkeonline.lintcode.common.ListNode;

/**
 * lintcode: #167, with Linked List
 */
public class AddTwoNumbers {
    public ListNode addLists(ListNode l1, ListNode l2) {

        ListNode dummy = new ListNode(0) ;
        ListNode curr = dummy ;

        int carry = 0;
        while(l1 != null && l2 != null) {
            int sum = (l1.val + l2.val + carry)   ;
            carry = sum / 10 ;
            ListNode node = new ListNode(sum % 10) ;
            curr.next = node ;
            curr = node ;
            l1 = l1.next ;
            l2 = l2.next ;
        }

        while(l1 != null) {
            int sum = (l1.val + carry) ;
            carry = sum / 10 ;
            ListNode node = new ListNode(sum % 10) ;
            curr.next = node ;
            curr = node ;
            l1 = l1.next ;
        }

        while(l2 != null) {
            int sum = (l2.val + carry)  ;
            carry = sum / 10 ;
            ListNode node = new ListNode(sum % 10) ;
            curr.next = node ;
            curr = node ;
            l2 = l2.next ;
        }

        if (carry != 0) {
            curr.next = new ListNode(carry);
        }

        return dummy.next ;
    }

}
